src/typechecker.erl

```diff
@@ -4109,36 +4110,44 @@ refine_ty(Ty1, Ty2, _, Env) ->
+                           Trace :: map(), Env :: env()}) ->
+          [gradualizer_type:abstract_type()].
+%% Produce the diff #{Assoc1} \ #{Assoc2}.
 %%
-%% M1 \ M2    | #{K := V} | #{K => V}
-%% -----------+-----------+--------------
-%% #{K := V}  | none      | #{K => V}
-%% #{K => V}  | #{}       | none
-%%
-%% Please note that we're changing field optionality, but not the types of keys or values.
-%%
-%% We might also see disjoint K1 from M1 and K2 from M2.
-%% In such a case we just leave K1 and its field unmodified.
-refine_map_field_ty({?type(map_field_exact, KVTy), ?type(map_field_exact, KVTy)}) ->
-    [];
-refine_map_field_ty({?type(map_field_assoc, KVTy), ?type(map_field_assoc, KVTy)}) ->
+%% Since Assoc2 is always computed from a pattern, it's always map_field_exact
+%% in practice.
+refine_map_field_ty({?type(map_field_exact, KVTy), ?type(map_field_exact, KVTy),
+                     _Trace, _Env}) ->
     [];
-refine_map_field_ty({?type(map_field_assoc, KVTy), ?type(map_field_exact, KVTy)}) ->
+refine_map_field_ty({?type(map_field_assoc, KVTy), ?type(map_field_exact, KVTy),
+                     _Trace, _Env}) ->
     %% #{x => y} \ #{x := y} = #{} -- i.e. an empty map
     [type(map, [])];
-refine_map_field_ty({?type(map_field_exact, KVTy), ?type(map_field_assoc, KVTy)}) ->
-    %% M1 = #{x := y}
-    %% M2 = #{x => y}
-    %% M1 \ M2 = #{x => y}
-    [type(map_field_assoc, KVTy)];
-refine_map_field_ty({?type(AssocTag1, _) = Assoc1, ?type(AssocTag2, _)})
-  when AssocTag1 == map_field_assoc, AssocTag2 == map_field_assoc;
-       AssocTag1 == map_field_exact, AssocTag2 == map_field_exact;
-       AssocTag1 == map_field_exact, AssocTag2 == map_field_assoc;
-       AssocTag1 == map_field_assoc, AssocTag2 == map_field_exact ->
-    %% TODO: this might lead to duplicates in the resulting type!
+refine_map_field_ty({?type(AssocTag1, [K1, V1]), ?type(map_field_exact, [K2, V2]),
+                     Trace, Env})
+  when AssocTag1 == map_field_assoc;
+       AssocTag1 == map_field_exact ->
+    %% This might lead to duplicates in the resulting type.
+    case refine(type(tuple, [K1, V1]), type(tuple, [K2, V2]), Trace, Env) of
+        ?type(none) ->
+            [];
+        ?type(tuple, [KeyRefined, ValueRefined]) ->
+            [type(AssocTag1, [KeyRefined, ValueRefined])];
+        ?type(union, Tuples) ->
+            %% TODO: Because we use the logic from tuples, this may produce
+            %% assocs with overlapping keys, e.g.
+            %%
+            %%     #{1..2 := 3..4} \ #{1 := 4} ===> #{2 := 3..4, 1..2 := 3}
+            %%
+            %% where we instead would like #{2 := 3..4, 1 := 3}, but another
+            %% problem is that only one key should be mandatory. Currently,
+            %% overlapping keys disable exhaustiveness checking.
+            [type(AssocTag1, [KeyRefined, ValueRefined])
+             || ?type(tuple, [KeyRefined, ValueRefined]) <- Tuples]
```

Comment:
I though about not doing any refinement in this case, if it confuses other code. I.e. just return [Assoc1] instead of this list comprehension.

Do you think it affects the self-check warning you saw?

Reply:
It's not related, no.